### ISSUE TITLE
Proposal for Reading Locabulary JSON from HTTP End-Point

### ISSUE THREAD

**Issue by jeremyf** [ndlib/locabulary#72]
On <span class="timestamp-wrapper"><span class="timestamp">&lt;2020-12-09 Wed&gt; </span></span> we [brainstormed options](https://docs.google.com/document/d/1iFg2RVuWlBaqTzY0rRApHkuogKnezhLthRHESc6AtQA/edit) for the [Locabulary gem](https://github.com/ndlib/locabulary).  The critical necessary change is that the product owners of CurateND want to push updates to the Administrative Units vocabulary for Sipity and CurateND.  At present, a change in data requires that we deploy the application.

## Background

Locabulary provides a set of `public @api` methods, all contained within the [top-level Locabulary module](https://github.com/ndlib/locabulary/blob/master/lib/locabulary.rb).  Both Sipity and CurateND use those public methods to populate their select boxes for creating and editing works.

## Proposal

Change `Locabulary::Utility.with_active_extraction_for` and `Locabulary::Utility.with_extraction_for` to read from an HTTP end-point instead of a local file system.  ~~My proposal would be to~~ We'll leverage Github's raw JSON URLs for these JSON objects.  That way, when we change the JSON via a PR, we have the change reflected in our HTTP source document.

The reader method would cache the results of the JSON body of each predicate.  This cache could be in memory or leverage a Rails cache store.  The easiest is to put it in memory and see how it behaves; This could be done with a class instance variable of the reader.

The reader method would treat a cached instance as stale after 4 hours.  *Note: I picked 4 hours to reflect that a half-day for an update is reasonable given the volume.  The product owners may determine a different cache expiry time frame.*

If we choose to use each Rails application’s cache, I believe we’d want to explore a configuration option for setting that cache store.  In other words, we’ll need to test Locabularly with a cache, and in unit tests I want a light weight cache (e.g. an in memory Hash).  We would want the test cache to conform to the Rails caching methods.

### Considerations for Failure

At present, Locabulary includes the data files.  In other words, if you have the gem on your machine, you have the data files available.  This means it is unlikely you’ll have an error reading a Locabulary JSON file. *If you did, you probably have a larger system-wide error happening.*

We would want to consider what to do if the remote HTTPS end-point is down.  I believe the best option would be to keep the cache valid (and bump its expiry by 2 hours).  This would highlight that we may want to use something more robust than a class instance variable.

From a design stand-point, I believe this highlights that the two applications should provide their cache; We’d want to ensure that each cache conforms to our expected interface.  We’d also then want Locabulary to create an in-memory implementation of that cache’s interface.

In the case where we have an application restart, and more specifically a clear cache AND Github's RAW content is down, then I propose that we start logging Sentry errors; And consider this an outage.  This, I believe will be the quickest and least full of antics option.  (Also, Github has a rather high uptime, so I think from a probability stand-point we are unlikely to have an empty cache AND see Github down).

**Comment by jeremyf** [comment]
For reference, the raw JSON for `./data/administrative_units.json` exists at https://github.com/ndlib/locabulary/raw/master/data/administrative_units.json

**Comment by dbrower** [comment]
@jeremyf Thank you for writing this up! I have a few comments.
1. I think the general direction is good. Let's make the proposal more concrete, and just say that it will pull the data from GitHub.
2. Not having a local copy would make important parts of both applications unusable. Instead of using "expire" language as though the local copy were a cache, maybe reframe as the local copy is a copy which is updated regularly from the source, but which doesn't necessarily "expire". 
3. I don't see a need to change the update interval in cases of retrieval errors. So long as everything is eventually updated, it should be good.
4. How would the case of GitHub being down and the applications restarting be handled?

**Comment by jeremyf** [comment]
@jeremyf Thank you for writing this up! I have a few comments.
> 1. I think the general direction is good. Let's make the proposal more concrete, and just say that it will pull the data from GitHub.

Agreed. 

> 2. Not having a local copy would make important parts of both applications unusable. Instead of using "expire" language as though the local copy were a cache, maybe reframe as the local copy is a copy which is updated regularly from the source, but which doesn't necessarily "expire". 

Let me rephrase here how I'm thinking about this.  Our preference is to treat the raw github JSON object as canonical.  We don't want to excessively hit that end-point (I don't know what that would mean).  So we'd be caching the JSON response document.

The local copy is an artifact of the reality that Locabulary packages up the JSON files as part of the gem distribution.  The local copy of each JSON file is (almost) guaranteed to have JSON data that is not any newer than what is up on RAW github.  It is in fact likely to have JSON data that is older than what's up on RAW github.  (e.g. we've merged in several administrative units but have not deploy locabularly changes).

> 3. I don't see a need to change the update interval in cases of retrieval errors. So long as everything is eventually updated, it should be good.

I think this relates to my thinking on caching the HTTP response; If our cache says its expired but the RAW end point is down, I think we may want to bump the expiry.  However, as I write this, I see that that adds another decision point in our logic.  So yes, we can instead say "If the cache has passed its point of expiration don't clear the cache until we have a successful request to the RAW end point.  Use that successful request to replace the cached data."

> 4. How would the case of GitHub being down and the applications restarting be handled?

Yeah…that sucks.  Here's a few considerations:

1) Start logging Sentry errors; And consider the site as down
2) Use the local file system's locabulary data (which would likely mean stale data); If we go this route, whenever we deploy, we'd want to always (as a matter of policy) upgrade the locabularly gem reference.  
3) As part of the caching process, store the JSON document locally and defer to that stored local copy.

I think we're looking at an infrequent scenario, so it could be something that could be a downtime event.

**Comment by dbrower** [comment]
> The local copy is an artifact of the reality that Locabulary packages up the JSON files as part of the gem distribution.

We have a terminology mismatch. I was using "local copy" to mean the most recent version of the file the application had, either in memory, or in a database. Not the copy coming from the locabulary gem.
 
> "If the cache has passed its point of expiration don't clear the cache until we have a successful request to the RAW end point. Use that successful request to replace the cached data."

Perfect!

As part of making the proposal more specific, could you choose one of the considerations for the restart-while-github-is-down and add it to the considerations for failure section?
